src/Microsoft.AspNetCore.Routing.Abstractions/EndpointMetadataCollection.cs

```diff
@@ -37,6 +41,7 @@ public EndpointMetadataCollection(IEnumerable<object> items)
             }
 
             _items = items.ToArray();
+            _cache = new ConcurrentDictionary<Type, object[]>();
```

Comment:
Have you considered populating a normal dictionary in the ctor instead of being lazy?

Edit:
As I typed this I realized we need to handle `GetMetadata<IInterface>()` which means reflection to populate the dictionary with all the different base types and interfaces of the values.

Reply:
I see your perf test will test this scenario but have we got any unit tests for it?

Reply:
Yeah, that was exactly my rationale as well, we don't know up front what `TMetadata` will be, so unless we wanted to use all possible types we couldn't do a good job.

Reply:
Yeah, so... oddly enough there are like no tests for EMC. I will remedy that.